// json-rpc-protocol/dist/format.spec.js
"use strict";
/* eslint-env jest */
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var errors_1 = require("./errors");
var format = __importStar(require("./format"));
var fromJson = JSON.parse;
// ===================================================================
describe('format', function () {
    it('.error()', function () {
        expect(fromJson(format.error(null, new errors_1.JsonRpcError('foo', 42)))).toEqual({
            error: {
                code: 42,
                message: 'foo'
            },
            id: null,
            jsonrpc: '2.0'
        });
    });
    it('.notification()', function () {
        expect(fromJson(format.notification('foo'))).toEqual({
            jsonrpc: '2.0',
            method: 'foo'
        });
    });
    it('.request()', function () {
        expect(fromJson(format.request(0, 'foo'))).toEqual({
            id: 0,
            jsonrpc: '2.0',
            method: 'foo'
        });
    });
    it('.response()', function () {
        expect(fromJson(format.response(1, 'foo'))).toEqual({
            id: 1,
            jsonrpc: '2.0',
            result: 'foo'
        });
    });
});
